### ISSUE TITLE
Phase 45.3: QuantizationManager — Mixed-Precision Quantization & Hardware Optimization

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#891]
### Overview
Implements comprehensive neural network quantization framework supporting post-training quantization (PTQ), quantization-aware training (QAT), mixed-precision quantization, and hardware-aware optimization. Reduces model size and inference latency by converting floating-point weights and activations to lower-precision representations.

#### 1. Post-Training Quantization (PTQ) (Nagel et al. 2021)
- Static quantization with calibration dataset
- Min-max, percentile, and entropy-based calibration
- Per-tensor and per-channel quantization granularity
- Cross-layer equalization (CLE) for improved PTQ accuracy
- Bias correction and AdaRound (Nagel et al. 2020)

#### 2. Quantization-Aware Training (QAT) (Jacob et al. 2018)
- Fake quantization operators in forward pass
- Straight-through estimator (STE) for gradient computation
- Learnable quantization parameters (scale, zero-point)
- Progressive QAT: start with higher precision, gradually reduce
- Batch normalization folding during quantization

#### 3. Mixed-Precision Quantization
- Per-layer bitwidth assignment (2/4/8/16/32-bit)
- Hardware-aware mixed-precision search (HAQ, Wang et al. 2019)
- Sensitivity-guided bitwidth allocation
- Pareto-optimal precision configurations
- HAWQ: Hessian-based bitwidth selection (Dong et al. 2019)

#### 4. Binary and Ternary Networks
- Binary weight networks (BWN): weights ∈ {-1, +1}
- XNOR-Net: binary weights and activations
- Ternary weight networks (TWN): weights ∈ {-1, 0, +1}
- Real-to-binary knowledge distillation
- Progressive binary training

#### 5. Hardware-Aware Quantization
- Target hardware latency models (CPU, GPU, NPU, DSP)
- Operator-level quantization support matrices
- INT8 optimization for ARM NEON / x86 AVX-512
- Quantization scheme selection per hardware target
- TensorRT/ONNX Runtime/CoreML quantization backends

### API Design

```python
class QuantizationManager:
    def __init__(self, config: QuantizationConfig):
        """Initialize quantization manager."""
    
    def quantize_ptq(
        self,
        model: nn.Module,
        calibration_loader: DataLoader,
        bitwidth: int = 8,
        granularity: str = "per_channel",  # per_tensor|per_channel
        calibration_method: str = "minmax"  # minmax|percentile|entropy
    ) -> QuantizedModel:
        """Apply post-training quantization with calibration."""
    
    def quantize_qat(
        self,
        model: nn.Module,
        train_loader: DataLoader,
        bitwidth: int = 8,
        epochs: int = 10
    ) -> QuantizedModel:
        """Train model with quantization-aware training."""
    
    def mixed_precision_search(
        self,
        model: nn.Module,
        val_loader: DataLoader,
        target_latency: float = None,
        target_size: float = None,
        hw_target: str = "generic"
    ) -> MixedPrecisionConfig:
        """Search for optimal per-layer bitwidth assignment."""
    
    def quantize(
        self,
        model: nn.Module,
        scheme: str = "ptq",  # ptq|qat|mixed|binary|ternary
        target_hw: str = "generic",
        **kwargs
    ) -> QuantizedModel:
        """Unified quantization interface."""
```

### Acceptance Criteria
- [ ] PTQ with min-max, percentile, and entropy calibration
- [ ] QAT with fake quantizers and STE
- [ ] Mixed-precision quantization with per-layer bitwidth search
- [ ] Binary and ternary network support
- [ ] Hardware-aware quantization for CPU/GPU/NPU targets
- [ ] INT8 models within 1% accuracy of FP32 on benchmarks
- [ ] 4x model size reduction with INT8 quantization
- [ ] Unit tests with >90% coverage

### References
- Jacob et al. (2018) "Quantization and Training of Neural Networks"
- Nagel et al. (2021) "A White Paper on Neural Network Quantization"
- Wang et al. (2019) "HAQ: Hardware-Aware Automated Quantization"
- Dong et al. (2019) "HAWQ: Hessian AWare Quantization"
- Rastegari et al. (2016) "XNOR-Net: ImageNet Classification Using Binary Convolutional Neural Networks"

> Part of [Phase 45 Planning](#888)
